lib/github-pages-health-check/domain.rb

```diff
@@ -0,0 +1,202 @@
+module GitHubPages
+  module HealthCheck
+    class Domain < GitHubPages::HealthCheck::Checkable
+
+      attr_reader :host
+
+      def initialize(host)
+        @host = host_from_uri(host)
+      end
+
+      # Runs all checks, raises an error if invalid
+      def check!
+        raise Errors::InvalidDNSError unless dns_resolves?
+        return true if proxied?
+        raise Errors::DeprecatedIPError if deprecated_ip?
+        raise Errors::InvalidARecord    if invalid_a_record?
+        raise Errors::InvalidCNAME      if invalid_cname?
+        raise Errors::NotServedByPages  unless served_by_pages?
+        true
+      end
+
+      def deprecated_ip?
+        return @deprecated_ip if defined? @deprecated_ip
+        @deprecated_ip = a_record? && old_ip_address?
```

Comment:
This leaves a bit of ambiguity and I'm not sure how Ruby's lexer deals with it. Parentheses around `a_record? && old_ip_address?` might help?